kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
+            case unifyData of
+                UnifyInjDirect _ -> decompose (injChild inj1) (injChild inj2)
+                UnifyInjSplit InjPair{inj1 = firstInj} ->
+                    if injFrom firstInj == injFrom inj1
+                        then decompose (injChild inj1) (evaluateInj inj2{injTo = injFrom inj1})
+                        else decompose (evaluateInj inj1{injTo = injFrom inj2}) (injChild inj2)
+                UnifyInjDistinct _ -> failMatch "distinct injections"
+        (Inj_ inj@Inj{injChild = App_ firstHead firstChildren}, secondTerm@(App_ secondHead _))
+            | Just unifyData <-
+                unifyOverloadingVsOverloaded
+                    overloadSimplifier
+                    secondHead
+                    secondTerm
+                    (Application firstHead firstChildren)
+                    inj{injChild = ()} ->
+                decomposeOverload $ flipResult unifyData
+        (firstTerm@(App_ firstHead _), Inj_ inj@Inj{injChild = App_ secondHead secondChildren})
+            | Just unifyData <-
+                unifyOverloadingVsOverloaded
+                    overloadSimplifier
+                    firstHead
+                    firstTerm
+                    (Application secondHead secondChildren)
+                    inj{injChild = ()} ->
+                decomposeOverload unifyData
+        (Inj_ inj1@Inj{injChild = App_ firstHead firstChildren}, Inj_ Inj{injChild = App_ secondHead secondChildren})
+            | Just unifyData <-
+                unifyOverloadingCommonOverload
+                    overloadSimplifier
+                    (Application firstHead firstChildren)
+                    (Application secondHead secondChildren)
+                    inj1{injChild = ()} ->
+                decomposeOverload unifyData
+        (_, _) | Just True <- List.isListSort tools sort ->
+            case (List.normalize pat, List.normalize subject) of
+                (Var_ var1, Var_ var2) | var1 == var2 -> discharge
+                (ElemVar_ var1, _) | isFunctionPattern subject -> bind (inject var1) subject
+                (SetVar_ var1, _) -> bind (inject var1) subject
+                (InternalList_ InternalList{internalListChild = l1}, InternalList_ InternalList{internalListChild = l2}) ->
+                    if length l1 == length l2
+                        then decomposeList $ zip (toList l1) (toList l2)
+                        else failMatch "list lengths are not equal"
+                (App_ symbol [InternalList_ InternalList{internalListChild = l1}, var@(ElemVar_ _)], InternalList_ InternalList{internalListChild = l2})
+                    | List.isSymbolConcat symbol ->
+                        if length l1 <= length l2
+                            then
+                                let l2' = Seq.drop (length l1) l2
+                                 in decomposeList $ (var, List.asInternal tools sort l2') : zip (toList l1) (toList (Seq.take (length l1) l2))
```

Comment:
Please factor this out and name it something suggestive.

Reply:
I'm not entirely sure what you're asking me to factor out here. As far as I can tell this entire code block is something that doesn't really repeat itself anywhere else. I don't think factoring out a function here would really make it more readable. But if you clarify what you're looking for here, I might be able to do something.